packages/assets-controllers/src/RatesController/RatesController.ts

```diff
@@ -0,0 +1,197 @@
+   * @param options.fetchMultiExchangeRate - Fetches the exchange rate from an external API. This option is primarily meant for use in unit tests.
+   * @param options.onStart - Optional callback to be executed when the polling stops.
+   * @param options.onStop - Optional callback to be executed when the polling starts.
+   */
+  constructor({
+    interval = 180000,
+    messenger,
+    state,
+    includeUsdRate,
+    fetchMultiExchangeRate = defaultFetchExchangeRate,
+    onStart,
+    onStop,
+  }: RatesControllerArgs) {
+    super({
+      name,
+      metadata,
+      messenger,
+      state: { ...defaultState, ...state },
+    });
+    this.#includeUsdRate = includeUsdRate;
+    this.#fetchMultiExchangeRate = fetchMultiExchangeRate;
+    this.#onStart = onStart;
+    this.#onStop = onStop;
+    this.#setIntervalLength(interval);
+  }
+
+  /**
+   * Sets the interval length for polling.
+   *
+   * @param intervalLength - The length of the interval in milliseconds.
+   */
+  #setIntervalLength(intervalLength: number) {
+    this.#intervalLength = intervalLength;
+  }
+
+  /**
+   * Updates the BTC rates by fetching new data.
+   */
+  async updateRates(): Promise<void> {
+    const releaseLock = await this.#mutex.acquire();
+    const { currency, cryptocurrencyList } = this.state;
+
+    try {
+      const response = await this.#fetchMultiExchangeRate(
+        currency,
+        cryptocurrencyList,
+        this.#includeUsdRate,
+      );
+
+      const updatedRates: ConversionRates = {};
+      for (const [key, value] of Object.entries(response)) {
+        updatedRates[key] = {
+          conversionDate: Date.now() / 1000,
+          conversionRate: value[currency],
+          usdConversionRate: value.usd || null,
+        };
+      }
+
+      this.update(() => {
+        return {
+          ...this.state,
+          rates: updatedRates,
+        };
+      });
+    } finally {
+      releaseLock();
+    }
+  }
+
+  /**
+   * Executes the polling operation to update rates.
+   */
+  async #executePoll(): Promise<void> {
+    await this.updateRates();
+  }
+
+  /**
+   * Starts the polling process.
+   */
+  async start(): Promise<void> {
+    if (this.#intervalId) {
+      return;
+    }
+
+    await this.#onStart?.();
+    this.#intervalId = setInterval(() => {
+      this.#executePoll().catch(console.error);
+    }, this.#intervalLength);
+  }
+
+  /**
+   * Stops the polling process.
+   */
+  async stop(): Promise<void> {
+    if (!this.#intervalId) {
+      return;
+    }
+
+    clearInterval(this.#intervalId);
+    this.#intervalId = undefined;
+    await this.#onStop?.();
+  }
+
+  getCryptocurrencyList(): string[] {
+    const { cryptocurrencyList } = this.state;
+    return cryptocurrencyList;
+  }
+
+  setCryptocurrencyList(list: string[]): void {
+    this.update(() => {
+      return {
+        ...this.state,
+        cryptocurrencyList: list,
+      };
+    });
+  }
+
+  async setCurrency(currency: string) {
```

Comment:
@mcmire how about `updateCurrencyAndRates`?

Reply:
Hmm, after looking at this some more, why does this method cause another fetch for rates, but `setCryptocurrencyList` does not? Both of them change a property in state that is read on each iteration of the polling loop anyway.

If the goal is to ensure that any time anything that the polling loop relies on is changed, that the polling loop is effectively restarted, perhaps there is another way to achieve that?

Reply:
Thats a good point. The reason for this method was to get the rates immediately to update the UI but I didn't realize it's the same case for that other method. Maybe encapsulating the logic can be an option?

I don't think it can't be done by listening to the StateChange event since the polling is also consistently changing the state.